README.md
# AdEx Validator Stack in Rust [![CI](https://github.com/AdExNetwork/adex-validator-stack-rust/workflows/Continuous%20Integration/badge.svg)](https://github.com/AdExNetwork/adex-validator-stack-rust/actions)

The Rust implementation of the Validator Stack

Reference implementation of the [AdEx validator stack](https://github.com/adexnetwork/adex-protocol#validator-stack-platform).

Components:

* [Sentry](#sentry)
* [Validator worker](#validator-worker)
* Adapter - Ethereum & Dummy (for testing) Adapters
* AdView manager

## Local & Testing setup

Requirements:

- Rust

  Check the [`rust-toolchain`](./rust-toolchain) file for specific version of rust.
  - [`cargo-make`](https://github.com/sagiegurari/cargo-make)
- Docker
- Node 14 (LTS) & npm
    Used for running a local Ethereum node with `ganache-cli` for automated tests in the `Ethereum Adapter` ([adapter/src/ethereum.rs](./adapter/src/ethereum.rs), also check [scripts/ethereum.sh](./scripts/ethereum.sh))

#### Linux

- `build-essentials` is required to build the project (error: `linker ``cc`` not found`)
- The crate `openssl-sys` requires `libssl-dev` and `pkg-config` for Ubuntu.

## Sentry

`Sentry` is the REST API that the [`Validator worker`](#validator-worker) uses for storing and retrieving information.
We need two services to be able to run `Sentry`: `Postgres` and `Redis`.

### Running Postgres

`docker run --rm --name adex-validator-postgres -e POSTGRES_PASSWORD=postgres -d -p 5432:5432 -v $HOME/docker/volumes/postgres:/var/lib/postgresql/data postgres`

- `$HOME/docker/volumes/postgres` - your local storage for postgres (persist the data when we remove the container)
- `POSTGRES_PASSWORD=postgres` - the password of `postgres` user

**NOTE:** Additionally you must setup 2 databases - `sentry_leader` & `sentry_follower` in order for the provided examples bellow to work.

### Running Redis

`docker run --rm -p 6379:6379 --name adex-validator-redis -d redis`

### Running Sentry Rest API

For a full list of all available CLI options on Sentry run `--help`:

```bash
cargo run -p sentry -- --help
```

Starting the Sentry API in will always run migrations, this will make sure the database is always up to date with the latest migrations, before starting and exposing the web server.

In `development` ( [`ENV` environment variable](#environment-variables) ) it will seed the database as well.

#### Using the `Ethereum Adapter`

The password for the Keystore file can be set using the [environment variable `KEYSTORE_PWD`](#adapter).

- Leader
    ```bash
    POSTGRES_DB="sentry_leader" PORT=8006 cargo run -p sentry -- \
        --adapter ethereum \
        --keystoreFile ./adapter/resources/keystore.json \
        ./docs/config/dev.toml
    ```

- Follower
    ```bash
    POSTGRES_DB="sentry_follower" PORT=8006 cargo run -p sentry -- \
        --adapter ethereum \
        --keystoreFile ./adapter/resources/keystore.json
        ./docs/config/dev.toml
    ```

#### Using the `Dummy Adapter`

**Dummy** identities:

- Leader: `ce07CbB7e054514D590a0262C93070D838bFBA2e`

```bash
    POSTGRES_DB="sentry_leader" PORT=8005 cargo run -p sentry -- \
        --adapter dummy \
        --dummyIdentity ce07CbB7e054514D590a0262C93070D838bFBA2e \
        ./docs/config/dev.toml
```
- Follower: `c91763d7f14ac5c5ddfbcd012e0d2a61ab9bded3`

```bash
    POSTGRES_DB="sentry_follower" PORT=8006 cargo run -p sentry -- \
        --adapter dummy \
        --dummyIdentity c91763d7f14ac5c5ddfbcd012e0d2a61ab9bded3 \
        ./docs/config/dev.toml
```

For full list, check out [primitives/src/util/tests/prep_db.rs#L29-L43](./primitives/src/util/tests/prep_db.rs#L29-L43)

#### Environment variables

- `ENV` - `production` or `development`; *default*: `development` - passing this env. variable will use the default configuration paths - [`docs/config/dev.toml`](./docs/config/dev.toml) (for `development`) or [`docs/config/prod.toml`](./docs/config/prod.toml) (for `production`). Otherwise you can pass your own configuration file path to the binary (check `cargo run -p sentry --help` for more information). In `development` it will make sure Sentry to seed the database.
- `PORT` - *default*: `8005` - The local port that Sentry API will be accessible at
- `IP_ADDR` - *default*: `127.0.0.1` - the IP address that the API should be listening to
- `ANALYTICS_RECORDER` - accepts any non-zero value - whether or not to start the `Analytics recorder` that will track analytics stats for payout events (`IMPRESSION` & `CLICK`)
##### Adapter
- `KEYSTORE_PWD` - Password for the `Keystore file`, only available when using `Ethereum Adapter` (`--adapter ethereum`)

##### Redis
- `REDIS_URL` - *default*: `redis://127.0.0.1:6379`

##### Postgres
- `POSTGRES_HOST` - *default*: `localhost`
- `POSTGRES_USER` - *default*: `postgres`
- `POSTGRES_PASSWORD` - *default*: `postgres`
- `POSTGRES_DB` - *default*: `user` name - Database name in Postgres to be used for this instance
- `POSTGRES_PORT` - *default*: `5432`

#####

### Running the Validator Worker

For a full list of all available CLI options on the Validator worker run `--help`:

```bash
cargo run -p validator_worker -- --help
```

#### Using the `Ethereum Adapter`
TODO: Update Keystore file and Keystore password for Leader/Follower as they are using the same at the moment.

The password for the Keystore file can be set using the environment variable `KEYSTORE_PWD`.

- Leader
    Assuming you have [Sentry API running](#running-sentry-rest-api) for the **Leader** on port `8005`:

    ```bash
    cargo run -p validator_worker
        --adapter ethereum
        --keystoreFile ./adapter/resources/keystore.json
        --sentryUrl http://127.0.0.1:8005
        ./docs/config/dev.toml
    ```

- Follower

    Assuming you have [Sentry API running](#running-sentry-rest-api) for the **Follower** on port `8006`:

    ```bash
    cargo run -p validator_worker
        --adapter ethereum
        --keystoreFile ./adapter/resources/keystore.json
        --sentryUrl http://127.0.0.1:8006
        ./docs/config/dev.toml
    ```

#### Using the `Dummy Adapter`
- Leader: `ce07CbB7e054514D590a0262C93070D838bFBA2e`

    Assuming you have [Sentry API running](#running-sentry-rest-api) for the **Leader** on port `8005`:

    ```bash
    cargo run -p validator_worker
        --adapter dummy
        --dummyIdentity ce07CbB7e054514D590a0262C93070D838bFBA2e
        --sentryUrl http://127.0.0.1:8005
        ./docs/config/dev.toml
    ```

- Follower: `c91763d7f14ac5c5ddfbcd012e0d2a61ab9bded3`

    Assuming you have [Sentry API running](#running-sentry-rest-api) for the **Follower** on port `8006`:

    ```bash
    cargo run -p validator_worker
        --adapter dummy
        --dummyIdentity c91763d7f14ac5c5ddfbcd012e0d2a61ab9bded3
        --sentryUrl http://127.0.0.1:8006
        ./docs/config/dev.toml
    ```

#### Environment variables

- `ENV`: `production` or `development` ( *default* ) - passing this env. variable will use the default configuration paths - [`docs/config/dev.toml`](./docs/config/dev.toml) (for `development`) or [`docs/config/prod.toml`](./docs/config/prod.toml) (for `production`). Otherwise you can pass your own configuration file path to the binary (check `cargo run -p sentry --help` for more information). In `development` it will make sure Sentry to seed the database.
- `PORT` - The local port that Sentry API will accessible at

##### Adapter
- `KEYSTORE_PWD` - Password for the `Keystore file`, only available when using `Ethereum Adapter` (`--adapter ethereum`)

## Development environment

We use [`cargo-make`](https://github.com/sagiegurari/cargo-make#overview) for running automated checks (tests, builds, formatting, code linting, etc.) and building the project locally
as well as on our Continuous Integration (CI). For a complete list of out-of-the-box commands you can check
[Makefile.stable.toml](https://github.com/sagiegurari/cargo-make/blob/master/src/lib/Makefile.stable.toml).

### Local development

Locally it's enough to ensure that the default development command is executing successfully:

```bash
cargo make
```

It will run `rustfmt` for you as well as `clippy` (it will fail on warnings) and it will run all the tests thanks to `cargo` (doc tests, unit tests, integration tests, etc.).

This will also run the [Automated tests](#automated-tests), so you must have `Redis` & `Postgres` running.

#### Automated tests

This requires [`cargo-make`](https://github.com/sagiegurari/cargo-make#overview) and since we have integration tests that require `Redis` ([see `Running Redis`](#running-redis)) & `Postgres` (see [`Running Postgres`](#running-postgres)), you need to be running those in order to run the automated tests:

`cargo make test`

You can relate to the [`Makefile.stable.toml`](https://github.com/sagiegurari/cargo-make/blob/master/src/lib/Makefile.stable.toml)
for more commands and cargo-make as a whole.


validator_worker/CHANGELOG.md
# Validator worker v0.2.0 (4 Jan 2021)

Fully featured Validator implementation written in Rust.


All the issues related to this release can be found here:

https://github.com/AdExNetwork/adex-validator-stack-rust/issues?q=is%3Aissue+is%3Aclosed+project%3AAdExNetwork%2Fadex-validator-stack-rust%2F2+milestone%3Avalidator-v0.2.0




sentry/src/db.rs
use bb8::Pool;
use bb8_postgres::tokio_postgres::NoTls;
use bb8_postgres::PostgresConnectionManager;
use redis::aio::MultiplexedConnection;
use redis::RedisError;
use std::env;

use lazy_static::lazy_static;

pub mod analytics;
mod channel;
pub mod event_aggregate;
mod validator_message;

pub use self::channel::*;
pub use self::event_aggregate::*;
pub use self::validator_message::*;

pub type DbPool = Pool<PostgresConnectionManager<NoTls>>;

lazy_static! {
    static ref REDIS_URL: String =
        env::var("REDIS_URL").unwrap_or_else(|_| String::from("redis://127.0.0.1:6379"));
    static ref POSTGRES_USER: String =
        env::var("POSTGRES_USER").unwrap_or_else(|_| String::from("postgres"));
    static ref POSTGRES_PASSWORD: String =
        env::var("POSTGRES_PASSWORD").unwrap_or_else(|_| String::from("postgres"));
    static ref POSTGRES_HOST: String =
        env::var("POSTGRES_HOST").unwrap_or_else(|_| String::from("localhost"));
    static ref POSTGRES_PORT: u16 = env::var("POSTGRES_PORT")
        .unwrap_or_else(|_| String::from("5432"))
        .parse()
        .unwrap();
    static ref POSTGRES_DB: Option<String> = env::var("POSTGRES_DB").ok();
}

pub async fn redis_connection() -> Result<MultiplexedConnection, RedisError> {
    let client = redis::Client::open(REDIS_URL.as_str())?;
    let (connection, driver) = client.get_multiplexed_async_connection().await?;
    tokio::spawn(driver);
    Ok(connection)
}

pub async fn postgres_connection() -> Result<DbPool, bb8_postgres::tokio_postgres::Error> {
    let mut config = bb8_postgres::tokio_postgres::Config::new();

    config
        .user(POSTGRES_USER.as_str())
        .password(POSTGRES_PASSWORD.as_str())
        .host(POSTGRES_HOST.as_str())
        .port(*POSTGRES_PORT);
    if let Some(db) = POSTGRES_DB.clone() {
        config.dbname(&db);
    }
    let pg_mgr = PostgresConnectionManager::new(config, NoTls);

    Pool::builder().build(pg_mgr).await
}

pub async fn setup_migrations(environment: &str) {
    use migrant_lib::{Config, Direction, Migrator, Settings};

    let settings = Settings::configure_postgres()
        .database_user(POSTGRES_USER.as_str())
        .database_password(POSTGRES_PASSWORD.as_str())
        .database_host(POSTGRES_HOST.as_str())
        .database_port(*POSTGRES_PORT)
        .database_name(&POSTGRES_DB.as_ref().unwrap_or(&POSTGRES_USER))
        .build()
        .expect("Should build migration settings");

    let mut config = Config::with_settings(&settings);
    config.setup().expect("Should setup Postgres connection");
    // Toggle setting so tags are validated in a cli compatible manner.
    // This needs to happen before any call to `Config::use_migrations` or `Config::reload`
    config.use_cli_compatible_tags(true);

    macro_rules! make_migration {
        ($tag:expr) => {
            migrant_lib::EmbeddedMigration::with_tag($tag)
                .up(include_str!(concat!("../migrations/", $tag, "/up.sql")))
                .down(include_str!(concat!("../migrations/", $tag, "/down.sql")))
                .boxed()
        };
    }

    let mut migrations = vec![
        make_migration!("20190806011140_initial-tables"),
        make_migration!("20200625092729_channel-targeting-rules"),
    ];

    if environment == "development" {
        // seeds database tables for testing
        migrations.push(make_migration!("20190806011140_initial-tables/seed"));
    }

    // Define Migrations
    config
        .use_migrations(&migrations)
        .expect("Loading migrations failed");

    // Reload config, ping the database for applied migrations
    let config = config.reload().expect("Should reload applied migrations");

    if environment == "development" {
        // delete all existing data to make tests reproducible
        Migrator::with_config(&config)
            .all(true)
            .direction(Direction::Down)
            .swallow_completion(true)
            .apply()
            .expect("Applying migrations failed");
    }

    let config = config.reload().expect("Should reload applied migrations");

    Migrator::with_config(&config)
        // set `swallow_completion` to `true`
        // so no error will be returned if all migrations have already been ran
        .swallow_completion(true)
        .show_output(true)
        .direction(Direction::Up)
        .all(true)
        .apply()
        .expect("Applying migrations failed");

    let _config = config
        .reload()
        .expect("Reloading config for migration failed");
}
